src/bridge/SequencerInbox.sol

```diff
@@ -307,37 +322,60 @@ contract SequencerInbox is GasRefundEnabled, ISequencerInbox {
         return (header, timeBounds);
     }
 
-    function formDataHash(bytes calldata data, uint256 afterDelayedMessagesRead)
+    function formEmptyDataHash(uint256 afterDelayedMessagesRead)
         internal
         view
-        validateBatchData(data)
         returns (bytes32, IBridge.TimeBounds memory)
     {
         (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
             afterDelayedMessagesRead
         );
-        bytes32 dataHash = keccak256(bytes.concat(header, data));
-        return (dataHash, timeBounds);
+        return (keccak256(header), timeBounds);
     }
 
-    function formEmptyDataHash(uint256 afterDelayedMessagesRead)
-        internal
-        view
-        returns (bytes32, IBridge.TimeBounds memory)
-    {
+    function formDataHash(
+        bytes calldata data,
+        uint256 afterDelayedMessagesRead,
+        IBridge.BatchDataLocation dataLocation
+    ) internal view returns (bytes32, IBridge.TimeBounds memory) {
+        uint256 fullDataLen = HEADER_LENGTH + data.length;
+        if (fullDataLen > maxDataSize) revert DataTooLarge(fullDataLen, maxDataSize);
+        if (data.length > 0 && (data[0] & DATA_AUTHENTICATED_FLAG) == DATA_AUTHENTICATED_FLAG) {
+            revert DataNotAuthenticated();
+        }
+
         (bytes memory header, IBridge.TimeBounds memory timeBounds) = packHeader(
             afterDelayedMessagesRead
         );
-        return (keccak256(header), timeBounds);
+        if (dataLocation == IBridge.BatchDataLocation.Blob) {
+            bytes32[] memory dataHashes = dataHashReader.getDataHashes();
+            if (dataHashes.length == 0) revert MissingDataHashes();
+            if (data.length != 1) revert InvalidBlobMetadata();
+            if (data[0] & DATA_BLOB_HEADER_FLAG == 0) revert InvalidBlobMetadata();
+
+            return (
+                keccak256(bytes.concat(header, data, abi.encodePacked(dataHashes))),
```

Comment:
There's a risk here of if the batch poster appends a fake 32 byte data hash to their data. Right now the system can't tell the difference between the batch poster's data and the sequencer inbox generated data hashes, since they're right next to each other and not length delimited.

Reply:
Perhaps I'm not understanding what you're saying, but the data is checked to be length 1 so that shouldnt be possible

Reply:
Ah yeah sorry I missed that on my phone. We could consider requiring the data just be the blob header byte. Any extra header info could be put inside the data.

Reply:
That is what we currently require, since we check length 1 and check the first byte is the blob header byte.

However I do agree that it's confusing to have the data here, when we're not really using it, I left it here for two reasons:
1. We may want to include extra 4844 related data in the future (maybe now?) such as the indexes of the blob versioned hashes we care about if the transaction sends to multiple entities.
2. It simplifies the internal representation to have a calldata object here

However, I think we can just do 1. if we need to, and 2. isnt important, so unless we think we want the granularity of the 4844 hash indices I'll remove it for now

Reply:
Regarding the extra data I agree it doesn't hurt to have that facility. For the specific case of blob indices, I don't think the submitter of the tx knows the blob indices for the blobs that will be in the slot for the block their tx gets included in, since at the very least there can be multiple blob bearing txes included in that block.

Reply:
The data block reader should only returns from this transaction, not from this block, which allows for index selection